simpleviewinc/keg-hub
# SectionList Component

**Tickets**
* [ZEN-498](https://jira.simpleviewtools.com/browse/ZEN-498)
* [ZEN-511](https://jira.simpleviewtools.com/browse/ZEN-511)

> **IMPORTANT**
> This PR should be tested along with [this PR](https://github.com/simpleviewinc/tap-events-force/pull/114)


## Updates - 1/21/2021
* added helper method to get scroll values `getScrollValues`
    * utilized this helper in `useScroll` method
    * we may want to get the scroll values outside of a hook
* `sectionList.native.js`
    * added logic to recalculate active session for this edge case:
        * if the section contents changes without scrolling, we want to make sure we update the activeSection accordingly

* added new `test` case in the `Testing` section of the [evf PR](https://github.com/simpleviewinc/tap-events-force/pull/114)

## Updates - 1/20/2021
- updated useScroll logic to account for when scrolling ends
- updated sectionList ref to be set the entire nativeObject ref rather than just a part of it
- updated `useSectionChangeOnScroll` to do calculations for new section only after scrolling ends

## Updates - 01/13/2021

* Fixed an issue with classNames being arrays vs a string
  * Updated StyleInjector to only join the class names to string when the component is a string
* Cleaned up useStyleCallback
  * Added defaults and extra memoize of styles object

<hr/>

[Semver-Status] minor

## Context

We need to update the Sessions Day navigation to match the updated mock-ups
See this ticket => https://jira.simpleviewtools.com/browse/ZEN-472
 

## Goal

* Utilize react-native sectionList
* On Desktop Web show gray line for section titles
* On Mobile show Day number for section titles
* Should also include the name of the day
* On day toggle, it should scroll to the section for that day
* On Desktop web, the day toggle day number needs to be updated when the gray line that separates days touches the bottom of the day toggle component
 

## Updates

* `repos/keg-components/configs/aliases.json`
  * Added alias for new components 
* `repos/keg-components/src/components/divider/divider.js`
 * Updated to use `forwardRef`
* `repos/keg-components/src/components/list/*`
  * Added `SectionList` component
  * Moved `useScrollClassNames` to it's own hook file so it can be reused between list components
  * Added some example of using the `SectionList` component
* `repos/keg-components/src/hooks/*`
  * Added two reusable hooks ( `useScrollClassNames` and `useScroll` )
  * Allows tracking browser scrolling in web environments
* `repos/keg-components/src/theme/components/list/*`
  * Updated / Added the styles for the `SectionList` component
* `repos/keg-components/src/utils/components/*`
  * `getElementLayout` - Helper for interact with an elements layout and position
  * `scrollList` - Helper for scrolling a list of items or the window

## Testing

* Setup
  * Pull this PR => `keg && keg pr 105`   

**Test 1**
* ~~Run the Keg-Components's docker package => `keg dpg run docker.pkg.github.com/simpleviewinc/keg-packages/keg-components:zen-498-session-list-view`~~
* Use this image instead => `keg consumer pack run docker.pkg.github.com/simpleviewinc/keg-packages/keg-test-consumer:master `
  * Navigate your browser to http://components-zen-498-session-list-view.local.kegdev.xyz/
  * In the Nav on the right, select `Components` => `List` => `SectionList`
  * Ensure the `SectionList` example displays properly
